File: src/main/java/com.foomoo.box/model/immutable/BoardModel.java
package com.foomoo.box.model.immutable;

import com.foomoo.box.*;
import com.foomoo.box.model.Vector;
import com.foomoo.box.model.Wall;

import java.util.*;
import java.util.Map.Entry;
import java.util.stream.IntStream;

public final class BoardModel {

    private final Player player;
    private final Map<Block, Cell> blockCellMap;
    private final Map<Target, Cell> targetCellMap;
    private final Cell minCell;
    private final Cell maxCell;

    private BoardModel(final BoardModelBuilder builder) {
        this.player = builder.player;
        this.blockCellMap = new HashMap<>(builder.blockCellMap);
        this.targetCellMap = new HashMap<>(builder.targetCellMap);

        this.minCell = builder.minCell;
        this.maxCell = builder.maxCell;
    }

    /**
     * Move the player to the given cell, pushing other blocks out of the way if necessary.
     *
     * @param to The cell to move the player to.
     * @return An optional of BoardModel specifying the new model following the movement of the player and any other
     * necessary pieces. The optional will be empty if the move could not be performed.
     */
    public Optional<BoardModel> movePlayerPieceTo(final Cell to) {
        BoardModelBuilder builder = new BoardModelBuilder(this);
        return recursiveBlockMove(builder, player, to, player.getPushStrength());
    }

    /**
     * Gets the cell for the given block.
     *
     * @param block The block to get the cell for.
     * @return Optional of the Cell if the block is present in the model, empty otherwise.
     */
    public Optional<Cell> getBlockCell(final Block block) {
        return Optional.ofNullable(blockCellMap.get(block));
    }

    /**
     * Gets the block at the given cell, if any.
     *
     * @param cell The cell to get the block for.
     * @return Optional of the Block at the cell. Empty if no block present.
     */
    public Optional<Block> getBlockAtCell(final Cell cell) {
        return blockCellMap.entrySet().stream().filter(entry -> Objects.equals(cell, entry.getValue())).map(Entry::getKey).findAny();
    }

    /**
     * Gets all blocks known to the model.
     *
     * @return The Set of Blocks.
     */
    public Set<Block> getBlocks() {
        return blockCellMap.keySet();
    }

    /**
     * Get the cell for the given target.
     *
     * @param target The target to get the cell for.
     * @return Optional of the Cell if the block is present in the model, empty otherwise.
     */
    public Optional<Cell> getTargetCell(final Target target) {
        return Optional.ofNullable(targetCellMap.get(target));
    }

    /**
     * Gets the target at the given cell, if any.
     *
     * @param cell The cell to get the target for.
     * @return Optional of Target at the cell. Empty if no target present.
     */
    public Optional<Target> getTargetAtCell(final Cell cell) {
        return targetCellMap.entrySet().stream().filter(entry -> Objects.equals(cell, entry.getValue())).map(Entry::getKey).findAny();
    }

    /**
     * Gets all targets known to the model.
     *
     * @return The Set of Targets.
     */
    public Set<Target> getTargets() {
        return targetCellMap.keySet();
    }

    public Cell getMinCell() {
        return minCell;
    }

    public Cell getMaxCell() {
        return maxCell;
    }

    public int getRowCount() {
        return maxCell.getRow() - minCell.getRow() + 1;
    }

    public int getColumnCount() {
        return maxCell.getColumn() - minCell.getColumn() + 1;
    }

    private Optional<BoardModel> moveBlocksBetweenCell(final BoardModelBuilder builder, final Block block, final Cell from, final Cell to, final int pushStrength) {

        // Assume the move will succeed and apply the change to the builder.
        builder.blockCell(block, to);

        // Which block is currently at the target.
        final Optional<Block> targetCellBlockOptional = getBlockAtCell(to);

        if (targetCellBlockOptional.isPresent()) {
            return targetCellBlockOptional.flatMap(targetCellBlock -> {
                // Get the push vector which can be applied to all blocks as necessarily.
                final Vector pushVector = to.subtract(from);

                if (pushStrength >= targetCellBlock.getEffortToMove()) {
                    final Vector translatedPushVector = targetCellBlock.translatePushVector(pushVector);
                    final Cell nextTargetCell = to.translate(translatedPushVector);

                    return moveBlocksBetweenCell(builder, targetCellBlock, to, nextTargetCell, pushStrength - targetCellBlock.getEffortToMove());
                } else {
                    return Optional.empty();
                }
            });
        } else {
            return Optional.ofNullable(builder.build());
        }
    }

    /**
     * Move the requested block, applying the change to the given builder. Push other blocks out of the way if required
     * as long as the number of blocks that can be pushed is not breached.
     *
     * @param builder      The builder to apply moved blocks to.
     * @param block        The block to move.
     * @param targetCell   The cell to move the block to.
     * @param pushStrength The number of blocks that can be pushed in the direction of movement to allow the given block
     *                     to move.
     * @return A new BoardModel specifying the board following all the piece movements. The model will be absent if the
     * move could not be performed.
     */
    private Optional<BoardModel> recursiveBlockMove(final BoardModelBuilder builder, final Block block, final Cell targetCell, final int pushStrength) {
        // Get the current cell for the block. If the block doesn't exist don't change the model.
        return Optional.ofNullable(blockCellMap.get(block)).flatMap(currentCell -> moveBlocksBetweenCell(builder, block, currentCell, targetCell, pushStrength));
    }


    public static class BoardModelBuilder {
        private final Player player;
        private final Map<Block, Cell> blockCellMap = new HashMap<>();
        private final Map<Target, Cell> targetCellMap = new HashMap<>();
        private Error error;
        private Cell minCell;
        private Cell maxCell;

        public BoardModelBuilder(final BoardModel originalModel) {
            this.player = originalModel.player;
            this.blockCellMap.putAll(originalModel.blockCellMap);
            this.targetCellMap.putAll(originalModel.targetCellMap);
        }

        public BoardModelBuilder(final Player player, final Cell playerCell) {
            this.player = player;
            blockCell(player, playerCell);
        }

        public BoardModelBuilder blockCell(final Block block, final Cell cell) {
            blockCellMap.put(block, cell);
            return this;
        }

        public BoardModelBuilder targetCell(final Target target, final Cell cell) {
            targetCellMap.put(target, cell);
            return this;
        }

        public BoardModelBuilder wall(final Cell cell) {
            blockCellMap.put(new Wall(), cell);
            return this;
        }

        public BoardModelBuilder wall(final Cell from, final Cell to) throws BoardModelException {
            // Only accept line walls in rows or columns.
            if (from.getRow() == to.getRow()) {
                IntStream.rangeClosed(from.getColumn(), to.getColumn()).mapToObj(column -> new Cell(from.getRow(), column)).forEach(this::wall);
            } else if (from.getColumn() == to.getColumn()) {
                IntStream.rangeClosed(from.getRow(), to.getRow()).mapToObj(row -> new Cell(row, from.getColumn())).forEach(this::wall);
            } else {
                throw new BoardModelException(String.format("Cannot create non-horizonal or non-vertical walls. Wall request from: %s, to: %s", from, to));
            }

            return this;
        }

        public BoardModel build() {
            blockCellMap.values().stream().reduce(Cell::minimalCell).ifPresent(cell -> minCell = cell);
            blockCellMap.values().stream().reduce(Cell::maximalCell).ifPresent(cell -> maxCell = cell);

            return new BoardModel(this);
        }
    }

}

File: src/main/scala/com/foomoo/box/model/immutable/BoardModel.scala
package com.foomoo.box.model.immutable

import com.foomoo.box._
import com.foomoo.box.model.Wall
import com.foomoo.box.model.immutable.BoardModel.BoardModelBuilder

import scala.collection.JavaConversions._

object BoardModel {

  class BoardModelBuilder(val player: Player, playerCell: Cell) {
    val blockCellMap: scala.collection.mutable.Map[Block, Cell] = scala.collection.mutable.Map(player -> playerCell)
    val targetCellMap: scala.collection.mutable.Map[Target, Cell] = scala.collection.mutable.Map()

    var error: Error = null
    var minCell: Cell = null
    var maxCell: Cell = null

    def this(originalModel: BoardModel) {
      this(originalModel.player, originalModel.blockCellMap.get(originalModel.player).get)
      blockCellMap.putAll(originalModel.blockCellMap)
      targetCellMap.putAll(originalModel.targetCellMap)
    }

    def blockCell(block: Block, cell: Cell): BoardModel.BoardModelBuilder = {
      blockCellMap.put(block, cell)
      this
    }

    def targetCell(target: Target, cell: Cell): BoardModel.BoardModelBuilder = {
      targetCellMap.put(target, cell)
      this
    }

    def wall(cell: Cell): BoardModel.BoardModelBuilder = {
      blockCellMap.put(new Wall, cell)
      this
    }

    def wall(fromCell: Cell, toCell: Cell): BoardModel.BoardModelBuilder = {
      val topLeft = Cell.minimalCell(fromCell, toCell)
      val bottomRight = Cell.maximalCell(fromCell, toCell)

      for (column <- topLeft.column to bottomRight.column) {
        wall(new Cell(topLeft.row, column))
        wall(new Cell(bottomRight.row, column))
      }

      for (row <- topLeft.row to bottomRight.row) {
        wall(new Cell(row, topLeft.column))
        wall(new Cell(row, bottomRight.column))
      }

      this
    }

    private def getAllCells = blockCellMap.values ++ targetCellMap.values

    def build: BoardModel = {
      minCell = getAllCells.reduce(Cell.minimalCell)
      maxCell = getAllCells.reduce(Cell.maximalCell)
      new BoardModel(this)
    }
  }

}

final class BoardModel(builder: BoardModelBuilder) {
  val player: Player = builder.player
  val blockCellMap: scala.collection.mutable.Map[Block, Cell] = builder.blockCellMap
  val targetCellMap: scala.collection.mutable.Map[Target, Cell] = builder.targetCellMap
  val minCell = builder.minCell
  val maxCell = builder.maxCell

  /**
    * Move the player to the given cell, pushing other blocks out of the way if necessary.
    *
    * @param to The cell to move the player to.
    * @return An Option of BoardModel specifying the new model following the movement of the player and any other
    *         necessary pieces. The option will be None if the move could not be performed.
    */
  def movePlayerPieceTo(to: Cell): Option[BoardModel] = {
    val builder: BoardModel.BoardModelBuilder = new BoardModel.BoardModelBuilder(this)
    recursiveBlockMove(builder, player, to, player.getPushStrength)
  }

  /**
    * Gets the cell for the given block.
    *
    * @param block The block to get the cell for.
    * @return Option of the Cell if the block is present in the model, empty otherwise.
    */
  def getBlockCell(block: Block): Option[Cell] = blockCellMap.get(block)

  /**
    * Gets the block at the given cell, if any.
    *
    * @param cell The cell to get the block for.
    * @return Option of the Block at the cell. Empty if no block present.
    */
  def getBlockAtCell(cell: Cell): Option[Block] = blockCellMap.find(_._2 == cell).map(_._1)

  /**
    * Gets all blocks known to the model.
    *
    * @return The Set of Blocks.
    */
  def getBlocks: Set[Block] = {
    blockCellMap.keySet.toSet
  }

  /**
    * Get the cell for the given target.
    *
    * @param target The target to get the cell for.
    * @return Optional of the Cell if the block is present in the model, empty otherwise.
    */
  def getTargetCell(target: Target): Option[Cell] = targetCellMap.get(target)

  /**
    * Gets the target at the given cell, if any.
    *
    * @param cell The cell to get the target for.
    * @return Option of Target at the cell. Empty if no target present.
    */
  def getTargetAtCell(cell: Cell): Option[Target] = targetCellMap.find(_._2 == cell).map(_._1)

  /**
    * Gets all targets known to the model.
    *
    * @return The Set of Targets.
    */
  def getTargets: Set[Target] = {
    targetCellMap.keySet.toSet
  }

  def getMinCell: Cell = {
    minCell
  }

  def getMaxCell: Cell = {
    maxCell
  }

  def getRowCount: Int = {
    maxCell.getRow - minCell.getRow + 1
  }

  def getColumnCount: Int = {
    maxCell.getColumn - minCell.getColumn + 1
  }

  private def moveBlocksBetweenCell(builder: BoardModel.BoardModelBuilder, block: Block, from: Cell, to: Cell, pushStrength: Int): Option[BoardModel] = {
    builder.blockCell(block, to)
    val targetCellBlockOptional: Option[Block] = getBlockAtCell(to)
    targetCellBlockOptional match {
      case None => Some(builder.build)
      case Some(targetCellBlock) =>
        // Get the push vector which can be applied to all blocks as necessarily.
        val pushVector = to.subtract(from)

        if (pushStrength >= targetCellBlock.getEffortToMove) {
          val translatedPushVector = targetCellBlock.translatePushVector(pushVector)
          val nextTargetCell = to.translate(translatedPushVector)
          moveBlocksBetweenCell(builder, targetCellBlock, to, nextTargetCell, pushStrength - targetCellBlock.getEffortToMove)
        } else {
          None
        }
    }
  }

  /**
    * Move the requested block, applying the change to the given builder. Push other blocks out of the way if required
    * as long as the number of blocks that can be pushed is not breached.
    *
    * @param builder      The builder to apply moved blocks to.
    * @param block        The block to move.
    * @param targetCell   The cell to move the block to.
    * @param pushStrength The number of blocks that can be pushed in the direction of movement to allow the given block
    *                     to move.
    * @return An Optional of BoardModel specifying the board following all the piece movements. The model will be None if the
    *         move could not be performed.
    */
  private def recursiveBlockMove(builder: BoardModel.BoardModelBuilder, block: Block, targetCell: Cell, pushStrength: Int): Option[BoardModel] = {
    blockCellMap.lift(block).flatMap(currentCell => moveBlocksBetweenCell(builder, block, currentCell, targetCell, pushStrength))
  }

}